### PR TITLE
SQL: eliminate INs and LIKEs where possibile

### PR DESCRIPTION
**Pull request by HansBull**
(in)equality operators cost less and should be preferred if subqueries return exactly one row.

### CODE REVIEW

File: src/common/image.c [hunk 1]
```diff
@@ -1680,8 +1680,8 @@ int32_t dt_image_rename(const int32_t imgid, const int32_t filmid, const gchar *
         (dt_database_get(darktable.db),
          "SELECT id"
          " FROM main.images"
-         " WHERE filename IN (SELECT filename FROM main.images WHERE id = ?1)"
-         "   AND film_id IN (SELECT film_id FROM main.images WHERE id = ?1)",
+         " WHERE filename = (SELECT filename FROM main.images WHERE id = ?1)"
```
Comment (TurboGit) [inline src/common/image.c:1683]: isn't this just equivalent to:

`WHERE filename = ?1 AND film_id = ?2`

Or I'm confused this morning ?

File: src/common/collection.c [hunk 2]
```diff
@@ -1458,7 +1458,7 @@ static gchar *get_query_string(const dt_collection_properties_t property, const
         query = dt_util_dstrcat(query, "(film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '%s%%'))",
                                 escaped_text);
       else
-        query = dt_util_dstrcat(query, "(film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '%s'))",
+        query = dt_util_dstrcat(query, "(film_id IN (SELECT id FROM main.film_rolls WHERE folder = '%s'))",
```
Comment (TurboGit) [inline src/common/collection.c:1461]: this will break a seach like `%e%` for example.

File: src/libs/collect.c [hunk 3]
```diff
@@ -1028,7 +1028,7 @@ static GtkTreeModel *_create_filtered_model(GtkTreeModel *model, dt_lib_collect_
         // Check if this path also matches a filmroll
         gtk_tree_model_get(model, &iter, DT_LIB_COLLECT_COL_PATH, &pth, -1);
         DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                                    "SELECT id FROM main.film_rolls WHERE folder LIKE ?1", -1, &stmt, NULL);
+                                    "SELECT id FROM main.film_rolls WHERE folder = ?1", -1, &stmt, NULL);
```
Comment (TurboGit) [inline src/libs/collect.c:1031]: again, won't this break seach using patterns ?

File: src/common/collection.c (hunk 2 above)
Comment (phweyland) [reply src/common/collection.c:1461]: not the same: the old one is case insensitive, not the second one.

File: src/common/image.c (hunk 1 above)
Comment (phweyland) [reply src/common/image.c:1683]: This is about duplicates. Searching for all images with the same name in the same folder.

The original query seems ok for  me.

File: src/common/image.c [hunk 4]
```diff
@@ -2304,7 +2304,7 @@ void dt_image_synch_all_xmp(const gchar *pathname)
       (dt_database_get(darktable.db),
        "SELECT id"
        " FROM main.images"
-       " WHERE film_id IN (SELECT id FROM main.film_rolls "
+       " WHERE film_id = (SELECT id FROM main.film_rolls "
```
Comment (phweyland) [inline src/common/image.c:2307]: This is not equivalent.
there may be several filmrolls for a single folder (multiple imports for images a of single folder), right ?

File: src/control/jobs/control_jobs.c [hunk 5]
```diff
@@ -960,8 +960,8 @@ static int32_t dt_control_delete_images_job_run(dt_job_t *job)
   free(imgs);
 
   DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                              "SELECT COUNT(*) FROM main.images WHERE filename IN (SELECT filename FROM "
-                              "main.images WHERE id = ?1) AND film_id IN (SELECT film_id FROM main.images WHERE "
+                              "SELECT COUNT(*) FROM main.images WHERE filename = (SELECT filename FROM "
+                              "main.images WHERE id = ?1) AND film_id = (SELECT film_id FROM main.images WHERE "
```
Comment (phweyland) [inline src/control/jobs/control_jobs.c:964]: same as above

File: src/libs/collect.c (hunk 3 above)
Comment (phweyland) [reply src/libs/collect.c:1031]: yes, plus case sensitivity ...

File: src/common/image.c (hunk 4 above)
Comment (TurboGit) [reply src/common/image.c:2307]: I don't think this is possible or maybe there is a feature that I have never seen. When you import a single image it is automatically placed into the folder corresponding to the directory the image is in. If you do that a second time for another image the second image will be placed into the same folder.

Comment (phweyland) [reply src/common/image.c:2307]: Just checked. You are right. film_id remains the same. My mistake.

Then that's ok for this one.

File: src/control/jobs/control_jobs.c (hunk 5 above)
Comment (phweyland) [reply src/control/jobs/control_jobs.c:964]: My mistake again. For the same reason as above, the change should be OK as well (the 2 internal select return a unique record).

**Comment by TurboGit** [comment]
@HansBull  : Any news about this one ?

**Comment by HansBull** [comment]
Yes and no. Inquiring how (much) those many many LIKEs would benefit from adding indices with COLLATE NOCASE.
I think I'll close this one and go back to the drawing board.

**Pull request closed without merging**